src/harness/testcases/WINNF_FT_S_SIQ_testcase.py

```diff
@@ -30,6 +30,60 @@ def tearDown(self):
     pass
 
   @winnforum_testcase
+  def test_WINNF_FT_S_SIQ_4(self):
+    """Tests related to DPA activated for some channels
+    
+    SAS sending responseCode = 0 successful response
+    and including all channels within frequency range FR.
+    """
+    # Trigger SAS to load DPAs
+    self._sas_admin.TriggerLoadDpas()
+    # Trigger SAS to de-active all the DPAs 
+    self._sas_admin.TriggerBulkDpaActivation({'activate':False})
+    # Trigger SAS to active one DPA on channel c
+    frequency_range = {
+      'lowFrequency': 3620000000.0,
+      'highFrequency': 3630000000.0
+    }
+    self._sas_admin.TriggerDpaActivation({'frequencyRange':frequency_range,\
+                                           'dpaId':'east_dpa4'})
+    time.sleep(300)
+    # Load and register CBSD
+    device_a = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_a.json')))
+    # Change device location to be inside DPA neighborhood
+    device_a['installationParam']['latitude'] = 30.71570
+    device_a['installationParam']['longitude'] = -88.09350
+    cbsd_ids = self.assertRegistered([device_a]) 
+    # Create and send spectrumInquiry with the same frequency range as DPA Channel
+    spectrum_inquiry_0 = json.load(
+        open(os.path.join('testcases', 'testdata', 'spectrum_inquiry_0.json')))
+    spectrum_inquiry_0['cbsdId'] = cbsd_ids[0]
+    spectrum_inquiry_0['inquiredSpectrum'][0]['lowFrequency'] = \
+                frequency_range['lowFrequency']
+    spectrum_inquiry_0['inquiredSpectrum'][0]['highFrequency'] = \
+                frequency_range['highFrequency']
+    request = {'spectrumInquiryRequest': [spectrum_inquiry_0]}
+    response = self._sas.SpectrumInquiry(request)['spectrumInquiryResponse'][0] 
+    # Check spectrum inquiry response
+    self.assertEqual(response['cbsdId'], cbsd_ids[0])
+    self.assertTrue('availableChannel' in response)
+    self.assertEqual(response['response']['responseCode'], 0)
+    # Verify available channels contains the requested range and don't have any missing or repeated channels
+    response['availableChannel'].sort(key=lambda ch: ch['frequencyRange']\
```

Comment:
Cute solution! It's probably worth bundling it into a utility function that (1) can be unit tested and (2) can be used by other test cases (I think SIQ.3 has a similar check).

As I understand it, this will sort available channels by lowFreq in reverse order, so we will see the high freq channels before the low freq channels. Given that, I think you need to swap the bodies of the 'if' statements below since the first element in the sorted list will be the highest freq channel. The body of the 'else' statement also needs a slight adjustment for the same reason, as right now it will trivially pass (instead I think you want to check that highFreq[i] >= lowFreq[i-1] -- but please double-check my logic).

Even with this adjustment, I think there's a corner case where, for example, we return [ [3565, 3575], [3550, 3560], [3550, 3565] ] as available channels (already sorted in this example), the inequality assertion will fail since 3560 is not >= 3565, even though the SAS has indeed provided information for [3550, 3575] (and the SAS would have passed if it had reversed the order of the last two elements). I think this could be fixed by sorting first by high freq then by low freq (assuming that the sort function preserves relative order of "equal" elements) but I'm not sure it's worth the hassle. If you agree with my assessment, consider adding a comment recognizing this limitation (and please pull out into its own function so that our implementations are consistent).

Reply:
thank you Kate for this useful comment, you are right I made an error, I wanted to sort them by lowFrequency, so the reverse should be set to false, I think by default it sorts the list then by high frequency(treat frequency range as tuple) but I added high frequency to the sort to be sure.
I fixed that and I added this function to the helper functions with unit tests in a different file.
I think the only assumption now is(which is not wrong assumption) for the response: the channel with min low frequency should have lowFreq exactly equals to requested low frequency and the channel with max high frequency should have high frequency equals to the request high frequency.

Reply:
[quick_testcase.py.txt](https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/files/1643870/quick_testcase.py.txt)

I did a quick test (see attached file; please excuse the ugly code and please remove the .txt extension that GitHub forced me to add) and this function seems to work as intended. We may need to add the ability to toggle the strict overlap requirement in order to use this with e.g. CBSD 2 in SIQ.3 (see the GAA channel check) but that could be done in a subsequent PR.

Reply:
I executed your tests and they pass, I don't know if you noticed a new file I added with the name : helper_unit_testcase.py, which does different test cases for this assert method. I thought it may be helpful to unit test some helper functions, but I think it shouldn't be executed with harness tests(product).

Reply:
Oh, excellent! Indeed I did not notice :) This is more than I was hoping for -- thank you! I added some small style comments but you can feel free to ignore them.